libs/vkd3d/device.c

```diff
@@ -4987,6 +4995,76 @@ struct d3d12_device *unsafe_impl_from_ID3D12Device(d3d12_device_iface *iface)
     return impl_from_ID3D12Device(iface);
 }
 
+static inline struct d3d12_device *d3d12_device_from_ID3D12DeviceExt(ID3D12DeviceExt *iface)
+{
+    struct d3d12_device_ext* d12device_ext = CONTAINING_RECORD(iface, struct d3d12_device_ext, ID3D12DeviceExt_iface);
+    return CONTAINING_RECORD(d12device_ext, struct d3d12_device, device_ext);
+}
+
+static ULONG STDMETHODCALLTYPE d3d12_device_ext_AddRef(ID3D12DeviceExt *iface)
+{
+    struct d3d12_device *device = d3d12_device_from_ID3D12DeviceExt(iface);
+    return ID3D12Device_AddRef(&device->ID3D12Device_iface);
+}
+
+static ULONG STDMETHODCALLTYPE d3d12_device_ext_Release(ID3D12DeviceExt *iface)
+{
+    struct d3d12_device *device = d3d12_device_from_ID3D12DeviceExt(iface);
+    return ID3D12Device_Release(&device->ID3D12Device_iface);
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_device_ext_QueryInterface(ID3D12DeviceExt *iface,
+        REFIID iid, void **out)
+{
+    TRACE("iface %p, iid %s, out %p.\n", iface, debugstr_guid(iid), out);
+
+    if (IsEqualGUID(iid, &IID_ID3D12DeviceExt) ||
+        IsEqualGUID(iid, &IID_IUnknown))
+    {
+        d3d12_device_ext_AddRef(iface);
+        *out = iface;
+        return S_OK;
+    }
+
+    WARN("%s not implemented, returning E_NOINTERFACE.\n", debugstr_guid(iid));
+
+    *out = NULL;
+    return E_NOINTERFACE;
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_device_ext_get_vulkan_handles(ID3D12DeviceExt *iface, void* vk_instance, void* vk_physical_device)
+{
+    struct d3d12_device *device = d3d12_device_from_ID3D12DeviceExt(iface);
+    *((VkInstance*)vk_instance) = device->vkd3d_instance->vk_instance;
+    *((VkPhysicalDevice*)vk_physical_device) = device->vk_physical_device;
+    return S_OK;
+}
+
+static BOOL STDMETHODCALLTYPE d3d12_device_get_extension_support(ID3D12DeviceExt *iface, D3D12_VK_EXTENSION  extension)
+{
+    // TODO: Use d3d12_device_CheckFeatureSupport() to query here.
+    //struct d3d12_device *device = d3d12_device_from_ID3D12DeviceExt(iface);
+    return true;
+}
+
+static CONST_VTBL struct ID3D12DeviceExtVtbl d3d12_device_ext_vtbl =
+{
+    /* IUnknown methods */
+    d3d12_device_ext_QueryInterface,
+    d3d12_device_ext_AddRef,
+    d3d12_device_ext_Release,
+
+    /* ID3D12DeviceExt methods */
+    d3d12_device_ext_get_vulkan_handles,
+    d3d12_device_get_extension_support
+};
+
+static HRESULT vkd3d12_device_ext_init(struct d3d12_device_ext* ext_device)
```

Comment:
Please can we be consistent with our naming here?

I'd also prefer it if we called stuff something like
`d3d12_device_vkd3d_ext` or something (same with the command list), just for clarity when reading that this isn't a part of D3D12.